rust/src/accountant.rs

```diff
@@ -0,0 +1,179 @@
+use crate::accumulator::{UsageAccumulator, UsageUnit};
+use crate::producer::{ClientError, KafkaConfig, KafkaProducer, Producer};
+use chrono::{Duration, Local};
+use serde_json::json;
+use std::ops::Drop;
+
+static DEFAULT_TOPIC_NAME: &str = "shared-resources-usage";
+/// This is the entry point for the library. It is in most cases
+/// everything you need to instrument your application.
+///
+/// The UsageAccountant needs a `Producer` and a `UsageAccumulator`.
+/// Data is stored in the accumulator and, periodically, it is
+/// flushed into the Kafka topic via a consumer.
+///
+/// Accumulating data locally is critical to reduce the performance impact
+/// of this library to a minimum and reduce the amount of Kafka messages.
+/// This means that this structure should be instantiated rarely.
+/// Possibly only once per application (or per thread).
+///
+/// Avoid creating a UsageAccountant every time some data needs to
+/// be recorded.
+
+pub struct UsageAccountant {
+    accumulator: UsageAccumulator,
+    producer: Box<dyn Producer>,
+    topic: String,
+}
+
+impl UsageAccountant {
+    /// Instantiates a UsageAccountant from a Kafka config object.
+    /// This initialization method lets the `UsageAccountant` create
+    /// the producer and own it.
+    ///
+    /// You should very rarely change topic name or granularity.
+    pub fn new(
+        producer_config: KafkaConfig,
+        topic_name: Option<&str>,
+        granularity: Option<Duration>,
+    ) -> UsageAccountant {
+        UsageAccountant::new_with_producer(
+            Box::new(KafkaProducer::new(producer_config)),
+            topic_name,
+            granularity,
+        )
+    }
+
+    /// Leaves the responsibility to provide a producer to the
+    /// client. Most of the times you should not need to use this.
+    pub fn new_with_producer(
+        producer: Box<dyn Producer>,
+        topic_name: Option<&str>,
+        granularity: Option<Duration>,
+    ) -> UsageAccountant {
+        let topic = topic_name.unwrap_or(DEFAULT_TOPIC_NAME).to_string();
+
+        UsageAccountant {
+            accumulator: UsageAccumulator::new(granularity),
+            producer,
+            topic,
+        }
+    }
+
+    /// Records an mount of usage for a resource, and app_feature.
+    ///
+    /// It flushes the batch if that is ready to be flushed.
+    /// The timestamp used is the system timestamp.
+    pub fn record(
+        &mut self,
+        resource_id: &str,
+        app_feature: &str,
+        amount: u64,
+        unit: UsageUnit,
+    ) -> Result<(), ClientError> {
+        let current_time = Local::now();
+        self.accumulator
+            .record(current_time, resource_id, app_feature, amount, unit);
+        if self.accumulator.should_flush(current_time) {
```

Comment:
Forgot this in the previous review. Please use monotonic time `Instant` instead of chrono durations for debouncing to receive a more stable interval.

Reply:
Are you sure? 
We do not only care about durations here. We also tie them to the timestamp to have a timestamp on the message we produce in Kafka.
My understanding of Instant is that it is meant to deal with monotonic durations not absolute timestamps.

Reply:
Anyway if I am wrong on this I will be happy to fix it in a following PR